File: src/tw/rmstudio/mt/java/course/mcu/ds/ex1/sorting/SortingAlgorithm.java
package tw.rmstudio.mt.java.course.mcu.ds.ex1.sorting;

public abstract class SortingAlgorithm {
    abstract int[] sort(int[] randomNumbers);

    public abstract String getAlgorithmName();

    abstract int getConditionCount();

    abstract int getSwapCount();

    public SortingResult sortResult(int[] randomNumbers) {
        long startTime, endTime;
        int[] sortedNumbers;

        startTime = System.nanoTime();
        sortedNumbers = sort(randomNumbers);
        endTime = System.nanoTime();

        return new SortingResult(getAlgorithmName(), sortedNumbers, endTime - startTime, getConditionCount(), getSwapCount());
    }
}

File: src/tw/rmstudio/mt/java/course/mcu/ds/ex1/sorting/SortingAlgorithm.kt
package tw.rmstudio.mt.java.course.mcu.ds.ex1.sorting

abstract class SortingAlgorithm {

    abstract val algorithmName: String

    internal abstract val conditionCount: Int

    internal abstract val swapCount: Int
    internal abstract fun sort(randomNumbers: IntArray): IntArray

    fun sortResult(randomNumbers: IntArray): SortingResult {
        val startTime = System.nanoTime()
        val sortedNumbers = sort(randomNumbers)
        val endTime = System.nanoTime()

        return SortingResult(algorithmName, sortedNumbers, endTime - startTime, conditionCount, swapCount)
    }
}
